### PR TITLE
Bindings onBeforeShapeIsolate?

### PR DESCRIPTION
**Pull request by ds300**
So we were kinda bending over backwards to capture the use case where we update the arrow's terminal x,y coords when unbinding, copy-pasting, and duplicating. 

- At first we abused the `onBeforeShapeDelete` callbacks, but that was footgunny.
- Then we created a `onBeforeUnbind` callback, which was less footgunny but still subtly footgunny.

This PR proposes reverting the `onBeforeUnbind` stuff, taking us back to having `onBeforeShapeDelete` stuff. But at the same time it adds `onBeforeShapeIsolate` callbacks which are triggered at the following times:

- When you delete the other shape in a bound shape pair
- When you copy/paste or duplicate one shape in a bound shape pair but not the other one
- When you opt-in while deleting bindings e.g. `deleteBindings([...], {isolateShapes: true})`

This PR also fixes the bound arrow drag interaction. We can probably extract that out to a separate PR if needed.

Kapture 2024-06-04 at 12 42 40

### Change Type

- [x] `sdk` — Changes the tldraw SDK
- [ ] `dotcom` — Changes the tldraw.com web app
- [ ] `docs` — Changes to the documentation, examples, or templates.
- [ ] `vs code` — Changes to the vscode plugin
- [ ] `internal` — Does not affect user-facing stuff

- [x] `bugfix` — Bug fix
- [ ] `feature` — New feature
- [x] `improvement` — Improving existing features
- [ ] `chore` — Updating dependencies, other boring stuff
- [ ] `galaxy brain` — Architectural changes
- [ ] `tests` — Changes to any test code
- [ ] `tools` — Changes to infrastructure, CI, internal scripts, debugging tools, etc.
- [ ] `dunno` — I don't know

### Test Plan

1. Add a step-by-step description of how to test your PR here.
2.

- [ ] Unit Tests
- [ ] End to end tests

### Release Notes

- Add a brief release note for your PR here.

### CODE REVIEW

File: packages/editor/src/lib/options.ts [hunk 1]
```diff
@@ -34,7 +34,7 @@ export interface TldrawOptions {
 	readonly collaboratorInactiveTimeoutMs: number
 	readonly collaboratorIdleTimeoutMs: number
 	readonly collaboratorCheckIntervalMs: number
-	readonly cameraMovingTimoutMs: number
```
Comment (SomeHats) [inline packages/editor/src/lib/options.ts:37]: why this change? should we apply this to the others?

File: packages/editor/src/lib/editor/Editor.ts [hunk 2]
```diff
@@ -8943,36 +8949,36 @@ function pushShapeWithDescendants(editor: Editor, id: TLShapeId, result: TLShape
...
+						bindingsWithBoth.add(binding.id)
+						continue
+					}
+					if (!hasFrom || !hasTo) {
+						bindingsToRemove.add(binding.id)
+					}
+				}
+			}
+
+			editor.deleteBindings([...bindingsToRemove], { isolateShapes: true })
```
Comment (SomeHats) [inline packages/editor/src/lib/editor/Editor.ts:8981]: thoughts on calling the onBeforeIsolate callbacks directly here rather than adding this as a special option to `deleteBindings`?

**Review by SomeHats — COMMENTED** [review]
looking good!

File: packages/editor/src/lib/options.ts (hunk 1 above)
Comment (ds300) [reply packages/editor/src/lib/options.ts:37]: just a driveby typo fix `Timout` -> `Timeout`

File: packages/editor/src/lib/editor/Editor.ts (hunk 2 above)
Comment (ds300) [reply packages/editor/src/lib/editor/Editor.ts:8981]: it doesn't work, the bindings need to actually be deleted in between calls to onBeforeIsolate otherwise the arrow geometries are incorrect when calculating the new bend.

File: packages/editor/src/lib/options.ts (hunk 1 above)
Comment (ds300) [reply packages/editor/src/lib/options.ts:37]: ah sorry seems i deleted the `Ms` bit by accident

File: packages/editor/src/lib/editor/Editor.ts [hunk 3]
```diff
@@ -5247,13 +5242,24 @@ export class Editor extends EventEmitter<TLEventMap> {
...
-		this.store.remove(ids)
+		if (isolateShapes) {
+			for (const id of ids) {
+				const binding = this.getBinding(id)
+				if (!binding) continue
+				const util = this.getBindingUtil(binding)
+				util.onBeforeIsolateFromShape?.({ binding, shape: this.getShape(binding.fromId)! })
+				util.onBeforeIsolateToShape?.({ binding, shape: this.getShape(binding.toId)! })
+				this.store.remove([id])
+			}
```
Comment (SomeHats) [inline packages/editor/src/lib/editor/Editor.ts:5255]: let's wrap this is `this.store.atomic` so that the `after` callbacks happen in the same way they do even without setting `isolateShapes`

**Review by SomeHats — APPROVED** [review]
lgtm

File: packages/editor/src/lib/editor/Editor.ts (hunk 3 above)
Comment (ds300) [reply packages/editor/src/lib/editor/Editor.ts:5255]: Hm that cause bugs if the binding's onAfterDelete does some important cleanup that the `onBeforeIsolate` callbacks need to function correctly. Is there any danger of something going wrong how it is?

Comment (SomeHats) [reply packages/editor/src/lib/editor/Editor.ts:5255]: If that's a source of bugs, then surely it would be a source of bugs for normal deletion too, right? When doing normal 1-sided deletion, these callbacks still get called, but the after-phase callbacks won't happen until everything has been deleted. I think you're more likely to end up with bugs as a result of two different calling patterns than a uniform one, considering that's how all the other after callbacks work.

Comment (ds300) [reply packages/editor/src/lib/editor/Editor.ts:5255]: huh yeah good point. alright I'll make this atomic and we can revisit the whole business if anybody complains.
